File: 0070-climbing-stairs/0070-climbing-stairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        if n == 1 or n == 2:
            return n
        a = 1
        b = 2
        c = 3

        for i in range(3, n+1):
            c = a + b
            temp_b = b
            b = c
            a = temp_b
        return c

# Space optimized bottom up DP Approach

# TC = O(n)
# SC = O(1)

File: 0070-climbing-stairs/0070-climbing-stairs.java
class Solution {
    public int climbStairs(int n) {
        int a = 1;
        int b = 2;
        int c = 3;

        if (n < 0) return -1;
        if (n == 0) return 1;
        if (n == 1 || n == 2) return n;

        for (int i = 3; i < n+1; i++) {
            c = a + b;
            a = b;
            b = c;
        }
        return b;
    }
}